snorkel/analysis/error_analysis.py

```diff
@@ -0,0 +1,207 @@
+from collections import Counter, defaultdict
+from typing import Any, List, Mapping, Optional, Sequence, Tuple
+
+import numpy as np
+
+from snorkel.types import ArrayLike
+
+from .utils import arraylike_to_numpy
+
+
+def error_buckets(
+    golds: ArrayLike, preds: ArrayLike, X: Optional[Sequence[Any]] = None
+) -> Mapping[Tuple[int, int], Any]:
+    """Returns examples (or their indices) bucketed by gold label/pred label combination
+
+    Parameters
+    ----------
+    gold
+        An ArrayLike of gold (int) labels
+    pred
+        An ArrayLike of (int) predictions
+    X
+        Optional, a sequence of examples corresponding to golds/preds
+        If not provided, indices will be returned instead
+
+    Returns
+    -------
+    Dict
+        A mapping of each error bucket to its corresponding indices/examples
+        If X is None, return indices
+            instead.
+
+    Returned buckets[i,j] is a list of items with predicted label i and true label j.
+    For a binary problem with (1=positive, 2=negative):
+        buckets[1,1] = true positives
+        buckets[1,2] = false positives
+        buckets[2,1] = false negatives
+        buckets[2,2] = true negatives
+    """
+    buckets: Mapping[Tuple[int, int], List[Any]] = defaultdict(list)
+    golds = arraylike_to_numpy(golds)
+    preds = arraylike_to_numpy(preds)
+    for i, (y, l) in enumerate(zip(preds, golds)):
+        buckets[y, l].append(X[i] if X is not None else i)
+    return dict(buckets)
+
+
+def confusion_matrix(
+    golds: ArrayLike,
+    preds: ArrayLike,
+    null_pred: bool = False,
+    null_gold: bool = False,
+    normalize: bool = False,
+    pretty_print: bool = True,
+) -> np.ndarray:
+    """Returns a confusion matrix for a set of golds/preds
+
+    Parameters
+    ----------
+    golds
+        an ArrayLike of gold (int) labels
+    preds
+        An Arraylike of (int) predictions
+    null_pred
+        If True, include the row corresponding to null predictions
+    null_gold
+        If True, include the col corresponding to null gold labels
+    normalize
+        If True, divide counts by the total number of items
+    pretty_print
+        If True, pretty-print the matrix before returning
+
+    Returns
+    -------
+    np.ndarray
+        A confusion matrix when mat[p, y] is the number of examples with prediction p
+        and true label y (following the typical convention).
+    """
+
+    conf = ConfusionMatrix(null_pred=null_pred, null_gold=null_gold)
+    golds = arraylike_to_numpy(golds)
+    preds = arraylike_to_numpy(preds)
+    conf.add(golds, preds)
+    mat = conf.compile()
+
+    if normalize:
+        mat = mat / len(golds)
+
+    if pretty_print:
+        conf.display(normalize=normalize)
+
+    return mat
+
+
+class ConfusionMatrix(object):
```

Comment:
I guess we've had this since metal days, but when do we use the iterative confusion matrix?

Reply:
True, we don't use the iterative building nature anymore, and it probably introduces unnecessary complexity to have it. I'll make a note to merge the class into the method down the road.